File: statemachine/src/main/java/org/nichel/sm/Status.java
package org.nichel.sm;

import android.app.Notification;
import android.app.NotificationManager;
import android.content.Context;
import android.os.Bundle;

import org.nichel.sm.annotation.OnEnterStatus;
import org.nichel.sm.annotation.OnExitStatus;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.ArrayList;

import timber.log.Timber;

public class Status {
    private final String label;

    private final Object target;
    private final Method[] methods;

    private final NotificationManager notificationManager;
    private static final int NOTIFICATION_ID = 0;
    private Notification notification;

    public Status(final Object target, final Context context, final String label) {
        this.label = label;
        this.target = target;
        this.methods = target.getClass().getDeclaredMethods();
        this.notificationManager = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);
        this.notification = null;
    }

    public String getLabel() {
        return label;
    }

    public void setNotification(final Notification notification) {
        this.notification = notification;
    }

    void enter(final StateMachine stateMachine, final Bundle bundle) throws IllegalAccessException, InvocationTargetException, InstantiationException {
        if (notification != null) {
            notificationManager.notify(NOTIFICATION_ID, notification);
        }

        onEnterStatus(stateMachine, this, bundle);
    }

    void exit(final StateMachine stateMachine) throws IllegalAccessException, InvocationTargetException, InstantiationException {
        if (notification != null) {
            notificationManager.cancel(NOTIFICATION_ID);
        }

        onExitStatus(stateMachine, this);
    }

    private void onEnterStatus(final StateMachine stateMachine, final Status status, final Bundle bundle) throws IllegalAccessException, InstantiationException, InvocationTargetException {
        for (final Method method : methods) {
            if (method.isAnnotationPresent(OnEnterStatus.class)) {
                final OnEnterStatus annotation = method.getAnnotation(OnEnterStatus.class);

                for (final String label : annotation.value()) {
                    if (label.equals(status.label)) {
                        final int params = method.getGenericParameterTypes().length;

                        if (params == 2) {
                            method.invoke(target, stateMachine, status);
                        } else if (params == 3) {
                            method.invoke(target, stateMachine, status, bundle);
                        }
                    }
                }
            }
        }
    }

    private void onExitStatus(final StateMachine stateMachine, final Status status) throws IllegalAccessException, InstantiationException, InvocationTargetException {
        for (final Method method : methods) {
            if (method.isAnnotationPresent(OnExitStatus.class)) {
                final OnExitStatus annotation = method.getAnnotation(OnExitStatus.class);

                for (final String label : annotation.value()) {
                    if (label.equals(status.label)) {
                        method.invoke(target, stateMachine, status);
                    }
                }
            }
        }
    }

    @Override
    public String toString() {
        return label;
    }
}

File: statemachine/src/main/java/org/nichel/sm/Status.kt
package org.nichel.sm

import org.nichel.sm.annotation.OnEnterStatus
import org.nichel.sm.annotation.OnExitStatus
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import java.util.*

public class Status(_target: Any, _label: String) {
    public val label = _label
    private val target = _target
    private val onEnterMethods = ArrayList<Method>()
    private val onExitMethods  = ArrayList<Method>()

    init {
        for (method in target.javaClass.declaredMethods) {
            if (method.isAnnotationPresent(OnEnterStatus::class.java)) {
                val annotation = method.getAnnotation(OnEnterStatus::class.java)

                for (value in annotation.value) {
                    if (value == label) {
                        onEnterMethods.add(method)
                    }
                }
            } else if (method.isAnnotationPresent(OnExitStatus::class.java)) {
                val annotation = method.getAnnotation(OnExitStatus::class.java)

                for (value in annotation.value) {
                    if (value == label) {
                        onExitMethods.add(method)
                    }
                }
            }
        }
    }

    @Throws(IllegalAccessException::class, InvocationTargetException::class)
    private fun onEnterStatus(stateMachine: StateMachine, status: Status, `object`: Any?) {
        for (method in onEnterMethods) {
            val params = method.genericParameterTypes

            if (params.size == 2) {
                method.invoke(target, stateMachine, status)
            } else if (params.size == 3) {
                method.invoke(target, stateMachine, status, `object`)
            }
        }
    }

    @Throws(IllegalAccessException::class, InvocationTargetException::class)
    private fun onExitStatus(stateMachine: StateMachine, status: Status) {
        for (method in onExitMethods) {
            method.invoke(target, stateMachine, status)
        }
    }

    @Throws(IllegalAccessException::class, InvocationTargetException::class)
    internal fun enter(stateMachine: StateMachine, `object`: Any?) {
        onEnterStatus(stateMachine, this, `object`)
    }

    @Throws(IllegalAccessException::class, InvocationTargetException::class)
    internal fun exit(stateMachine: StateMachine) {
        onExitStatus(stateMachine, this)
    }
}